### PR TITLE
Warn when font isn't found and another is chosen

### LINKED ISSUE #1017: Display a dialog when the user selects an invalid font
# Summary of the new feature/enhancement

Currently when changing the 'fontFace' to a font that doesn't exist the font size just gets so small (~0.5 font size) that you can't read the text.

# Proposed technical implementation details (optional)

It would be nice if the terminal informed the user that the font is not valid - Maybe a pop-up/modal?

### PR DESCRIPTION
**Pull request by miniksa**
Display a warning message when the DirectX renderer resolves a font that isn't the one you selected to warn that it couldn't be found.

## PR Checklist
* [x] Closes #1017
* [x] I work here
* [x] Tested manually.
* [x] Am core contributor. We talked about it.

## Detailed Description of the Pull Request / Additional comments

image

Also I wrote the dialog event chain out of `TermControl` to be reusable in the future for other messages the control might want to tell a host about and various levels. 

## Validation Steps Performed

- Manual validation, setting bad font name, fixing font name with `settings.json`.

### CODE REVIEW

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 1]
```diff
@@ -1928,6 +1930,48 @@ namespace winrt::TerminalApp::implementation
...
+            break;
+        case TerminalControl::NoticeLevel::Error:
+            title = RS_(L"NoticeError"); //\xe783
+            break;
+        }
+
+        _ShowControlNoticeDialog(title, message);
+    }
+
+    void TerminalPage::_ShowControlNoticeDialog(winrt::hstring title, winrt::hstring message)
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1958]: consider `const &` for the params

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -1979,14 +1979,35 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+            // Pop the rest of this function to the tail of the UI thread
+            // Just in case someone was holding a lock when they called us and
+            // the handlers decide to do something that take another lock
+            // (like ShellExecute pumping our messaging thread...GH#7994)
+            co_await Dispatcher();
+
+            auto noticeArgs = winrt::make_self<NoticeEventArgs>(NoticeLevel::Warning, message);
+            _raiseNoticeHandlers(*strongThis, *noticeArgs);
+        }
+
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:2010]: be very careful here! this co_await will change the thread that fontSizeChangedHandlers is called on later.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 3]
```diff
@@ -1979,14 +1979,35 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+        // If the actual font isn't what was requested...
+        if (_actualFont.GetFaceName() != _desiredFont.GetFaceName())
+        {
+            // Then warn the user that we picked something because we couldn't find their font.
+
+            // Save things we need to resume later.
+            auto strongThis{ get_strong() };
+
+            // Format message with user's choice of font and the font that was chosen instead.
+            const winrt::hstring message { fmt::format(std::wstring_view(RS_(L"NoticeFontNotFound")), _desiredFont.GetFaceName(), _actualFont.GetFaceName()) };
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:1999]: Consider using `wstring_view{ }` instead of `()`. I'm trying to discourage the use of a constructor as a function call :smile:

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw [hunk 4]
```diff
@@ -223,12 +223,12 @@
...
-    <value>&#x2022; Found a keybinding with too many strings for the "keys" array. There should only be one string value in the "keys" array.</value>
-    <comment>{Locked="\"keys\"","&#x2022;"} This glyph is a bullet, used in a bulleted list.</comment>
+    <value>• Found a keybinding with too many strings for the "keys" array. There should only be one string value in the "keys" array.</value>
+    <comment>{Locked="\"keys\"","•"} This glyph is a bullet, used in a bulleted list.</comment>
   </data>
   <data name="MissingRequiredParameter" xml:space="preserve">
-    <value>&#x2022; Found a keybinding that was missing a required parameter value. This keybinding will be ignored.</value>
-    <comment>{Locked="&#x2022;"} This glyph is a bullet, used in a bulleted list.</comment>
+    <value>• Found a keybinding that was missing a required parameter value. This keybinding will be ignored.</value>
+    <comment>{Locked="•"} This glyph is a bullet, used in a bulleted list.</comment>
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Resources/en-US/Resources.resw:231]: undo this change plz

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 5]
```diff
@@ -1928,6 +1930,48 @@ namespace winrt::TerminalApp::implementation
...
+            break;
+        case TerminalControl::NoticeLevel::Info:
+            title = RS_(L"NoticeInfo"); // \xe946
+            break;
+        case TerminalControl::NoticeLevel::Warning:
+            title = RS_(L"NoticeWarning"); //\xe7ba
+            break;
+        case TerminalControl::NoticeLevel::Error:
+            title = RS_(L"NoticeError"); //\xe783
+            break;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1952]: What are all of these `//\xebe8` for?

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 6]
```diff
@@ -1928,6 +1930,48 @@ namespace winrt::TerminalApp::implementation
...
+
+        _ShowControlNoticeDialog(title, message);
+    }
+
+    void TerminalPage::_ShowControlNoticeDialog(winrt::hstring title, winrt::hstring message)
+    {
+        if (auto presenter{ _dialogPresenter.get() })
+        {
+            // FindName needs to be called first to actually load the xaml object
+            auto controlNoticeDialog = FindName(L"ControlNoticeDialog").try_as<WUX::Controls::ContentDialog>();
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1963]: Wait...Really? So you never actually use `controlNoticeDialog`? I guess might as well not set it to the variable (or call `Title` off of `controlNoticeDialog` below).

File: src/cascadia/TerminalApp/TerminalPage.xaml [hunk 7]
```diff
@@ -69,6 +69,16 @@ the MIT License. See LICENSE in the project root for license information. -->
             DefaultButton="Primary">
         </ContentDialog>
 
+        <ContentDialog
+            x:Load="False"
+            x:Name="ControlNoticeDialog"
+            x:Uid="ControlNoticeDialog"
+            DefaultButton="Primary">
+            <TextBlock IsTextSelectionEnabled="True" TextWrapping="WrapWholeWords">
+                <Run x:Name="NoticeReason"/>
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.xaml:78]:
```suggestion
                <Run x:Name="NoticeMessage"/>
```
idk

Comment (carlos-zamora) [reply src/cascadia/TerminalApp/TerminalPage.xaml:78]: Don't really care about this one.

**Review by carlos-zamora — COMMENTED** [review]
Most of these things are pretty small.

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 2 above)
Comment (miniksa) [reply src/cascadia/TerminalControl/TermControl.cpp:2010]: oof, didn't mean to do that. per our teams msg, use lambda trick to throw it into space.

```
[captures]() -> winrt::fire_and_forget {
    co_await Dispatcher();
    _DOTHETHINGHandlers();
}();
```

**Comment by mdtauk** [comment]
Could this warning also be displayed inline with the new Settings UI?  For when loading a the JSON, and the font name is not recognised or no longer installed.

Not sure if validation will be ready to apply to the ComboBox / DropDownBox

**Comment by DHowett** [comment]
No, for the settings UI we're actually going to force you to toggle the bits of the font name in by clicking on a button at 9600 baud and give you no feedback whatsoever.

OF COURSE the settings UI is going to tell if you if your font isn't valid

**Comment by carlos-zamora** [comment]
_Ideally,_ we could make it a dropdown! Took a quick look at FluentTerminal and they did something interesting. They listed all of your fonts, but made the monospace ones bold.

**Comment by DHowett** [comment]
This is not the appropriate place to have this discussion.

**Comment by mdtauk** [comment]
> No, for the settings UI we're actually going to force you to toggle the bits of the font name in by clicking on a button at 9600 baud and give you no feedback whatsoever.
> 
> OF COURSE the settings UI is going to tell if you if your font isn't valid

Well that is good then.  The image shows a dialog, so I just wanted to bring it up.

The Settings UI itself, should only list the compatible fonts (whether it is Monospaced fonts only, or some more technical qualifier), which makes it less likely to happen (unless it uses an Editable ComboBox and an invalid font name is typed in).

But as the settings can be edited outside of the app, means when loading the app and it's settings - a check should be made for a valid font, a message displayed either on load as a ContentDialog, as well as in the Settings UI - with a fallback (probably Cascadia)

File: src/cascadia/TerminalControl/TermControl.idl [hunk 8]
```diff
@@ -45,6 +45,20 @@ namespace Microsoft.Terminal.TerminalControl
         String Uri { get; };
     }
 
+    enum NoticeLevel
+    {
+        Debug = 10,
+        Info = 20,
+        Warning = 30,
+        Error = 40,
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.idl:53]: Do these values have a particular meaning?

**Review by zadjii-msft — COMMENTED** [review]
I don't think I have much that people haven't already mentioned

File: src/cascadia/TerminalControl/TermControl.idl (hunk 8 above)
Comment (miniksa) [reply src/cascadia/TerminalControl/TermControl.idl:53]: No, I just pulled 4 levels out of my butt and left space for future ones to be put between without renumbering, like BASIC used to do...

If you have another idea... lemme know.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 6 above)
Comment (miniksa) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1963]: Well.... the trickery is, as far as I know, that these things are lazy-instantiated. So I believe this is necessary to make XAML load the object before I can do the things below. So yeah, I don't use it. But I had to ask for it to make sure XAML did the needful first on the relevant other variables.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 5 above)
Comment (miniksa) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1952]: I had this bright idea that I was going to use Segoe icons for a bug shaped icon for debug and an I in a circle info icon for informational. But it didn't work right because the text in the dialog won't font fallback. And in lieu of adding another text field with the font set just right to take only the Segoe icon.... I just left the codepoints here for if I cared enough in the future because it was mildly painful to look them up in the first place.

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw (hunk 4 above)
Comment (miniksa) [reply src/cascadia/TerminalApp/Resources/en-US/Resources.resw:231]: wtf how did that happen.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 1 above)
Comment (miniksa) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1958]: sold.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 5 above)
Comment (DHowett) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1952]: Yea les keep em

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 9]
```diff
@@ -1987,6 +1987,27 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+
+            // Capture what we need to resume later.
+            [strongThis = get_strong(), message]() -> winrt::fire_and_forget {
+                // Pop the rest of this function to the tail of the UI thread
+                // Just in case someone was holding a lock when they called us and
+                // the handlers decide to do something that take another lock
+                // (like ShellExecute pumping our messaging thread...GH#7994)
+                co_await strongThis->Dispatcher();
+
+                auto noticeArgs = winrt::make_self<NoticeEventArgs>(NoticeLevel::Warning, message);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:2006]: Tiny nit:

You can use use `make` which returns the projected type directly. You can then std::move that into the constructor for your args type (and dispense with the `*`)

Doing it with the "self" variant introduces an intermediate `com_ptr` to the implementation type itself.

File: src/cascadia/TerminalControl/TermControl.h [hunk 10]
```diff
@@ -81,6 +82,22 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         hstring _uri;
     };
 
+    struct NoticeEventArgs :
+        public NoticeEventArgsT<NoticeEventArgs>
+    {
+    public:
+        NoticeEventArgs(NoticeLevel level, hstring message) :
+            _level(level),
+            _message(message) {}
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.h:91]: Nit: you can either use `const&` for message or `std::move()` it on initialization to save a refcount.

**Review by DHowett — APPROVED** [review]

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw (hunk 4 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/Resources/en-US/Resources.resw]: The VS resw editor just reloads the whole file and always replaced these. So used to seeing that haha.

File: .github/actions/spell-check/expect/expect.txt [hunk 11]
```diff
@@ -2828,4 +2828,4 @@ zy
 AAAAABBBBBBCCC
 AAAAA
 BBBBBCCC
-abcd
\ No newline at end of file
+abcd
```
Comment (carlos-zamora) [inline .github/actions/spell-check/expect/expect.txt]: huh, another change that isn't really a change...

**Review by carlos-zamora — APPROVED** [review]
Idk what the deal with those two "changes, but not changes" are, but meh.

File: .github/actions/spell-check/expect/5757ec679b03a4240130c3c53766c91bbc5cd6a7.txt [hunk 12]
```diff
@@ -1 +1 @@
-renamer
\ No newline at end of file
+renamer
```
[the remark this answers is not in the document]
Comment (DHowett) [reply .github/actions/spell-check/expect/5757ec679b03a4240130c3c53766c91bbc5cd6a7.txt]: End of file newline changes to calm the linters

File: .github/actions/spell-check/expect/expect.txt (hunk 11 above)
Comment (miniksa) [reply .github/actions/spell-check/expect/expect.txt]: its a whitespace EOL. ignore it. it was warning in the spell checker.

File: .github/actions/spell-check/expect/5757ec679b03a4240130c3c53766c91bbc5cd6a7.txt (hunk 12 above)
[the remark this answers is not in the document]
Comment (miniksa) [reply .github/actions/spell-check/expect/5757ec679b03a4240130c3c53766c91bbc5cd6a7.txt]: had to add a newline at EOF to shut up the spell checker.

**Review by DHowett — APPROVED** [review]
